### ISSUE TITLE
Multiple Constraints with different k and maximum value of k

### ISSUE THREAD

**Issue by LeoFHall** [openmm/openmm#4492]
Hi, I am quite new to openmm so sorry if this is a dumb question, I tried looking at previous issues and documentation but I have a problem with my simulation.
I'm trying to simulate a MeOH crystal structure with centre of mass motion, one particle being fixed in place and the others having a spring constant that I'll graually decrease (in seperate simulations not in one)

If I try to constrain by setting
`system.setParticleMass(FixedParticle, 0*amu)
`
I get:
`Exception: A constraint cannot involve a massless particle
`
Even if the Restraint isn't set (I assume it's because of the bonds to the neighbouring Carbon and Hydrogen, I'm trying to pin in place an Oxygen)
If I try to have two constraints though they interact in a weird way and the Constrained atoms wobbles around and interestingly, it wobbles around more if k is either low, which makes sense, but it also gets worse if k is very high (10e15)

Thanks in advance for your help !
Leo

Here is the full code for when I have two constraints :
```
# Input Files
pdb = PDBFile('Methanol-conf.pdb')
forcefield = ForceField('OPLS-AA_toy-MeOH.xml')

# System Configuration
nonbondedMethod = PME 
nonbondedCutoff = 1*nanometers
ewaldErrorTolerance = 0.00001 
constraints = None 
constraintTolerance = 0.000001
vdwCutoff = 1*nanometer
LambdaE = 6666.67 #Eintein Field spring constant
#Vega MeOH done with LambdaE = 6666.67 at 150K
kspring = 16#62800*kilojoules_per_mole/nanometer 
fixSpring = 1000000000*kilojoules_per_mole/nanometer
FixedParticle = 1 #index of the constrained atom, first oxygen is second particle, indexes start at 0

# Integration Options
dt = 0.0005*picoseconds
temperature = 150*kelvin #temp at which the the crystal structure was taken
friction = 1.0/picosecond
pressure = 1.0*bar   #NpT so pressure is specified
barostatInterval = 25
collisionFreq = 1/picosecond

# Simulation Options
steps = 2000 #just 2000 for test run, all the reporters are divided by 100 too
equilibrationSteps = 1000
pdbReporter = PDBReporter('trajectory-toyMeOH-DA1-ConStrain.pdb', 100)
dataReporter = StateDataReporter('log-toyMeOH-DA1-ConStrain.csv', 100, totalSteps=steps,\
    step=True, speed=True, progress=True, potentialEnergy=True, temperature=True, volume=True, totalEnergy=True)

# Prepare the Simulation
print('Building system...')
topology = pdb.topology
positions = pdb.positions


#topology = modeller.topology
#positions=modeller.positions


#Creating the system
#removeCMMotion=False to allow the centre of mass to drift
system = forcefield.createSystem(topology, nonbondedMethod=nonbondedMethod, nonbondedCutoff=nonbondedCutoff, vdwCutoff=vdwCutoff,removeCMMotion=False)

#Adding a force to restrain atoms
restraint = CustomExternalForce('k*periodicdistance(x,y,z,x0,y0,z0)^2')
restraint.addGlobalParameter('k',kspring)
restraint.addPerParticleParameter('x0') #for each particle a reference position is determined
restraint.addPerParticleParameter('y0')
restraint.addPerParticleParameter('z0')
for atom in topology.atoms(): #Apply restraint to all the particles except the fixed one
    if atom.index != FixedParticle:
        restraint.addParticle(atom.index,pdb.positions[atom.index]) #pdb.positions[i] gives the desired reference position ?
system.addForce(restraint)

#Adding a constraint to the frozen reference atom by giving a massive harmonic constraint
constraint = CustomExternalForce('k*periodicdistance(x,y,z,x0,y0,z0)^2')
constraint.addGlobalParameter('k',fixSpring)
constraint.addPerParticleParameter('x0') 
constraint.addPerParticleParameter('y0')
constraint.addPerParticleParameter('z0')

constraint.addParticle(FixedParticle,pdb.positions[FixedParticle]) 
system.addForce(constraint)

#system.setParticleMass(FixedParticle, 0*amu) #commented out here but that's how I fix it when Constraint isn't applied


system.addForce(AndersenThermostat(temperature, collisionFreq)) #Andersen thermostat to add temperature coupling to the simulation
system.addForce(MonteCarloAnisotropicBarostat((1,1,1)*bar, temperature,True, True, True)) #Add barostat for the constant pressure parametre
#anisotropic to allow for the box to change shape
integrator = VerletIntegrator(dt)
integrator.setConstraintTolerance(constraintTolerance)
print("chk3")

simulation = Simulation(topology, system, integrator)#, platform)
print("chk4")

simulation.context.setPositions(positions)

# Minimize and Equilibrate
print('Performing energy minimization...')
simulation.minimizeEnergy()
print('Equilibrating...')
simulation.context.setVelocitiesToTemperature(temperature)
simulation.step(equilibrationSteps)

# Simulate
print('Simulating...')
simulation.reporters.append(pdbReporter)
simulation.reporters.append(dataReporter)
#simulation.reporters.append(chkptReporter)
simulation.currentStep = 0
simulation.step(steps)
```

**Comment by peastman** [comment]
Let's first clarify the terminology a bit.

A *constraint* is something you add to the System by calling `addConstraint()`.  It sets a mathematical condition that must be exactly satisfied by the integrator.  If you constrain two particles to remain a fixed distance apart, the simulation will only ever produce conformations in which they're that distance apart.

A *restraint* is a force you add to the system.  It encourages the simulation to remain close to satisfying a certain condition by increasing the energy when it gets too far away.

Your CustomExternalForce is a restraint, not a constraint.  That isn't what the error message is talking about.

Does your system contain any water?  That's the only way I can see you would have any constraints.  You don't call `addConstraint()`, and you don't include the `constraints` argument to `createSystem()`.  But if you have any water molecules, they'll have constraints added to make them rigid unless you specify `rigidWater=False`.

**Comment by LeoFHall** [comment]
Right, when I say constraint I mean that I don't want it to move, but given the mass=0 method you suggest in here:
https://gist.github.com/peastman/21facc27bed0ee2437a2011dbf67aab9
gave an error, I "Constrained" it by having a restraint with a very high spring constant, which is where I observed a weird behaviour.

There are no waters in the system just 300 methanol molecules
Thanks for your swift anwser !

**Comment by peastman** [comment]
Try adding this code to your script.

```python
print(system.getNumConstraints(), 'constraints')
for i in range(system.getNumConstraints()):
    print(system.getConstraintParameters(i))
```

That will tell you exactly what constraints you have.

**Comment by peastman** [comment]
Are you sure you didn't pass `constraints=Hbonds` when calling `createSystem()`?  I don't see any other way they would have gotten there.

Anyway, the error message means exactly what it says: a constraint cannot involve a massless particle.  If you want to make a particle massless, it can't be part of a constraint.

**Comment by LeoFHall** [comment]
I'm sure, all my code is in the first message the only lines I didn't paste were commented out. The constraints shown are those from the forcefield I think. (I'm using OPLS-AA but modified because it's a MeOH toy model where the CH3 is one sphere).

I know, which is why I was trying to simulate it as a particle with a very strong restrain instead, but there's an issue when I have two restraints where the second one doesn't come into play it seems

**Comment by LeoFHall** [comment]
that and when the spring constant I define is very high, the particle wobbles more. could it be that when it's too high it's not dealt well with by python ?

**Comment by LeoFHall** [comment]
Oh sorry I didn't see the latest message yesterday, here is the python code, pdb input file, and the .xml file for the OPLS-AA forcefield adapted to the methanol toy model
[toy-MeOH-DeltaA1.py.txt](https://github.com/openmm/openmm/file
[Methanol-conf.pdb.txt](https://github.com/openmm/openmm/files/14838046/Methanol-conf.pdb.txt)
s/14838037/toy-MeOH-DeltaA1.py.txt)
[OPLS-AA_toy-MeOH.xml.txt](https://github.com/openmm/openmm/files/14838065/OPLS-AA_toy-MeOH.xml.txt)

**Comment by peastman** [comment]
The correct PDB identifier for methanol is [MOH](http://ligand-expo.rcsb.org/reports/M/MOH/index.html).  You incorrectly call it SOL, which makes OpenMM think it's water.  (The correct identifier for water is HOH, but SOL is a commonly encountered incorrect one some programs use for it, so OpenMM assumes that's what you mean.)  And since the default is `rigidWater=True`, it constrains the bonds to make it rigid.

**Comment by LeoFHall** [comment]
Oh I would've though "SOL" being defined in the forcefield file would've trumped that, thanks for the help I'll modify the pdb and try with that !

**Comment by peastman** [comment]
[This FAQ entry](https://github.com/openmm/openmm/wiki/Frequently-Asked-Questions#template) may help to explain how it works.  The name used in the force field doesn't matter at all.  It's just an arbitrary label.

**Comment by LeoFHall** [comment]
oh ok !
but does that mean that my toy methanol model with CH3 being a single sphere isn't going to be useable, since it won't correspond to the expected MeOH ?

**Comment by peastman** [comment]
It's ok to use a coarse grained force field with implicit hydrogens.  Just make sure what's in your PDB file matches the template in the force field.

**Comment by LeoFHall** [comment]
Ok great, thanks for the help !
